### PR TITLE
mgr/dashboard: List subvolume groups

### PR DESCRIPTION
**Pull request by Pegonzal**
Added listing for subvolume groups.

I've worked on top of #52526, so that PR should be merged first.

image

Fixes: https://tracker.ceph.com/issues/62168
Signed-off-by: Pedro Gonzalez Gomez <pegonzal@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/pybind/mgr/dashboard/controllers/cephfs.py [hunk 1]
```diff
@@ -564,3 +565,18 @@ def ls_dir(self, fs_id, path=None, depth=1):
...
+class CephFSSubvolumeGroups(RESTController):
+    def __init__(self):  # pragma: no cover
+        super().__init__()
+
+    def get(self, vol_name):
+        error_code, out, err = mgr.remote('volumes', '_cmd_fs_subvolumegroup_ls', None, {'vol_name': vol_name})
+        if error_code != 0:
+            raise RuntimeError(
+                f'Error listing subvolume groups')
+        return json.loads(out)
```
Comment (cloudbehl) [inline src/pybind/mgr/dashboard/controllers/cephfs.py:582]: Do we add some doc text for our swagger API? So to tell what needs to be passed to the API endpoint to get output.

Also. when we don't pass the vol_name, what will happen? Can it fetch us all the volume in that case?

**Comment by cloudbehl** [comment]
As part of the listing page, I think these are set of parameters which could add value.

- Subvolume name
- Volume name(Required when we will have separate listing page, to identify which volume this subvolume belongs too)
- State
- Type
- Data pool
- Mode of FS
- Path
- Bytes used
- Bytes quota(if set)
- Created_at

**Comment by epuertat** [comment]
I'm not familiar with subvolume groups, but here it goes my comments regarding the tab ordering:
- I'd keep the Performance tab as the last (for consistency; it's like the default place in wherever "Performance" tabs appear).
- From a hierarchical perspective, subvolumes lie above or beneath directories?

**Comment by epuertat** [comment]
BTW, so far it doesn't seem to bring much info, just the subvolume group name... Maybe we could put the subvolume groups inside the main table as a set of labels/tags column?

File: src/pybind/mgr/dashboard/controllers/cephfs.py [hunk 2]
```diff
@@ -583,3 +584,31 @@ def ls_dir(self, fs_id, path=None, depth=1):
...
         return paths
+
+
+
+@APIRouter('/cephfs/subvolume/group', Scope.CEPHFS)
+@APIDoc("Cephfs Subvolume Group Management API", "Cephfs")
+class CephFSSubvolumeGroups(RESTController):
+    def __init__(self):  # pragma: no cover
+        super().__init__()
+
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/controllers/cephfs.py:595]: do we need this?

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cephfs/cephfs-form/cephfs-form.component.spec.ts [hunk 3]
```diff
@@ -1,2 +1 @@
-describe('CephfsVolumeFormComponent', () => {
-});
+describe('CephfsVolumeFormComponent', () => {});
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cephfs/cephfs-form/cephfs-form.component.spec.ts:1]: spillover from @pereman2's commit?

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cephfs/cephfs-subvolume-group/cephfs-subvolume-group.component.html [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+</ng-template>
+
+<ng-template #modeToHumanReadableTpl
+             let-value="value">
+  <span *ngFor="let result of (value | octalToHumanReadable)"
+        [ngClass]="result.class"
+        [ngbTooltip]="result.toolTip">
+    {{ result.content }}
+  </span>
+</ng-template>
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cephfs/cephfs-subvolume-group/cephfs-subvolume-group.component.html:43]: I think we can generalize octal-to-human-conversion in tables component and create a `cellTransformation` and use it on both place.

**Review by nizamial09 — APPROVED** [review]
LGTM! Thanks @Pegonzal probably we can make this PR dependent on my Subvolumes PR so some of the changes won't get duplicated!

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cephfs/cephfs-subvolume-group/cephfs-subvolume-group.component.ts [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+  subvolumeGroup$: Observable<CephfsSubvolumeGroup[]>;
+
+  constructor(private cephfsSubvolumeGroup: CephfsSubvolumeGroupService) {}
+
+  ngOnInit(): void {
+    this.columns = [
+      {
+        name: $localize`Name`,
+        prop: 'name',
+        flexGrow: 0.6
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cephfs/cephfs-subvolume-group/cephfs-subvolume-group.component.ts:46]: Let's make the `name` bold!
```suggestion
        flexGrow: 0.6
        cellTransformation: CellTemplate.bold
```

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cephfs/cephfs-subvolume-group/cephfs-subvolume-group.component.ts [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+        name: $localize`Created`,
+        prop: 'info.created_at',
+        flexGrow: 0.5,
+        cellTransformation: CellTemplate.timeAgo
+      }
+    ];
+  }
+
+  ngOnChanges() {
+    this.subvolumeGroup$ = this.cephfsSubvolumeGroup.get(this.fsName);
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cephfs/cephfs-subvolume-group/cephfs-subvolume-group.component.ts:80]: Let's handle the error as well
```suggestion
    this.subvolumeGroup$ = this.cephfsSubvolumeGroup.get(this.fsName).pipe(
    catchError(() => {
        this.context.error();
        return of(null);
      })
   );
```

File: src/pybind/mgr/dashboard/controllers/cephfs.py [hunk 7]
```diff
@@ -37,6 +39,24 @@ def list(self):
         fsmap = mgr.get("fs_map")
         return fsmap['filesystems']
 
+    def create(self, name: str, service_spec: Dict[str, Any]):
+        service_spec_str = '1 '
```
Comment (avanthakkar) [inline src/pybind/mgr/dashboard/controllers/cephfs.py:43]: Can be made much simpler with `.join()`
```suggestion
        labels_str = ','.join(f'label:{label}' for label in placement.get('labels', []))
        hosts_str = ','.join(placement.get('hosts', []))
       
       service_spec_str = f"1 {labels_str}{',' if labels_str and hosts_str else ''}{hosts_str}"
```

File: src/pybind/mgr/dashboard/controllers/cephfs.py [hunk 8]
```diff
@@ -37,6 +39,24 @@ def list(self):
         fsmap = mgr.get("fs_map")
         return fsmap['filesystems']
 
+    def create(self, name: str, service_spec: Dict[str, Any]):
```
Comment (avanthakkar) [inline src/pybind/mgr/dashboard/controllers/cephfs.py:42]: Can you add the API tests for this endpoint and  a unittest for cephfs service in frontend as well?

**Comment by nizamial09** [comment]
@Pegonzal there are still some conflicts!

File: src/pybind/mgr/dashboard/controllers/cephfs.py [hunk 9]
```diff
@@ -654,3 +655,27 @@ def get(self, vol_name: str):
...
+            error_code, out, err = mgr.remote('volumes', '_cmd_fs_subvolumegroup_ls', None, {'vol_name': vol_name})
+            if error_code != 0:
+                raise RuntimeError(
+                    f'Error listing subvolume groups')
+            subvolume_groups =  json.loads(out)
+            for group in subvolume_groups:
+                error_code, out, err = mgr.remote('volumes', '_cmd_fs_subvolumegroup_info', None, {'vol_name': vol_name,
+                'group_name': group['name']})
+                if error_code != 0:
+                    raise RuntimeError(
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/controllers/cephfs.py:674]:
```suggestion
                    raise DashboardException(
```

File: src/pybind/mgr/dashboard/controllers/cephfs.py [hunk 10]
```diff
@@ -654,3 +655,27 @@ def get(self, vol_name: str):
...
+                        f'Failed to get info for subvolume group {group["name"]}: {err}'
+                    )
+                group['info'] = json.loads(out)
+            return subvolume_groups
+        else:
+            raise RuntimeError(
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/controllers/cephfs.py:680]:
```suggestion
            raise DashboardException(
```

File: src/pybind/mgr/dashboard/controllers/cephfs.py [hunk 11]
```diff
@@ -654,3 +655,27 @@ def get(self, vol_name: str):
                 )
             subvolume['info'] = json.loads(out)
         return subvolumes
+
+@APIRouter('/cephfs/subvolume/group', Scope.CEPHFS)
+@APIDoc("Cephfs Subvolume Group Management API", "Cephfs")
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/controllers/cephfs.py:660]:
```suggestion
@APIDoc("Cephfs Subvolume Group Management API", "CephfsSubvolumeGroup")
```

**Comment by nizamial09** [comment]
jenkins retest this please

File: src/pybind/mgr/dashboard/controllers/cephfs.py [hunk 12]
```diff
@@ -674,3 +674,29 @@ def create(self, vol_name: str, subvol_name: str, **kwargs):
...
+
+    def get(self, vol_name):
+        if not vol_name:
+            raise DashboardException(
+                'Error listing subvolume groups')
+        error_code, out, err = mgr.remote('volumes', '_cmd_fs_subvolumegroup_ls',
+                                          None, {'vol_name': vol_name})
+        if error_code != 0:
+            raise DashboardException(
+                'Error listing subvolume groups')
```
Comment (pereman2) [inline src/pybind/mgr/dashboard/controllers/cephfs.py:691]: nit: would be nice to have the vol name in conjunction with the message

**Review by pereman2 — APPROVED** [review]

**Comment by nizamial09** [comment]
looks like unrelated e2e failure. I remember fixing it up before. Will check it again!

```
Login Page
    1) "after all" hook for "types login credentials and takes screenshot"


  0 passing (17s)
  1 failing

  1) Login Page
       "after all" hook for "types login credentials and takes screenshot":
     Error: Eyes-Cypress detected diffs or errors during execution of visual tests. See details at: https://eyes.applitools.com/app/batches/00000251710683509698/00000251710683509571?accountId=hCTedwnSTUWSCZlXPqua0g~~
       Passed - 2 tests
         ✓ Login Component Check [1920x1080]
         ✓ Login Component Check [1920x1080]
       Diffs detected - 2 tests
         ⚠ Dashboard Component [1920x1080]
         ⚠ Dashboard Component [1920x1080]

       See details at: https://eyes.applitools.com/app/batches/00000251710683509698/00000251710683509571?accountId=hCTedwnSTUWSCZlXPqua0g~~
```

File: src/pybind/mgr/dashboard/controllers/cephfs.py (hunk 12 above)
Comment (nizamial09) [reply src/pybind/mgr/dashboard/controllers/cephfs.py:691]: yup, that'd be good. Let's take it along with the [create subvolume group ](https://github.com/ceph/ceph/pull/52886/ )PR @Pegonzal . its hard to get a passing CI now.

**Pull request merged**